Add EVPN and VTEP API

tssurya opened the pull request:
## 📑 Description

This PR implements:

1. EVPN API fields in CUDN
2. VTEP CRD
3. NetConf interface fields (i.e translation from CUDN) 
4. NetInfo() struct changes to access these fields

Depends on https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5089 and https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5644/

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [x] My code requires changes to the documentation
- [x] if so, I have updated the documentation as required
- [x] My code requires tests
- [x] if so, I have added and/or updated the tests as required
- [x] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it

E2Es were added, and manually just create a cluster and try to create:

```
apiVersion: k8s.ovn.org/v1
kind: VTEP
metadata:
  name: vtep-dualstack
spec:
  cidrs:
  - "100.67.0.0/24"
  - "fd00:100:67::/64"
```
OR
```
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: l2-evpn-primary
spec:
  namespaceSelector: {matchLabels: {kubernetes.io/metadata.name: red}}
  network:
    topology: Layer2
    layer2:
      role: Primary
      subnets: ["10.20.100.0/24"]
    transport: EVPN
    evpnConfiguration:
      vtep: evpn-vtep
      macVRF:
        vni: 100
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * EVPN transport support with VRF configuration and VTEP resource; VTEP CRD now installed by default.

* **Documentation**
  * VTEP and EVPN API reference pages added and cross‑linked.

* **API Changes**
  * Transport enum now includes EVPN.
  * Field renamed: spec.noOverlayOptions → spec.noOverlay (NoOverlayConfig).

* **Tests**
  * New unit and e2e EVPN/VTEP test suites and scenarios added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tssurya commented:
```
+ true
+ kubectl apply -f k8s.ovn.org_vteps.yaml
The CustomResourceDefinition "vteps.k8s.ovn.org" is invalid: 
* spec.validation.openAPIV3Schema.properties[status].properties[allocations].items.properties[ip].x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
* spec.validation.openAPIV3Schema.properties[status].properties[allocations].items.properties[ip].x-kubernetes-validations[0].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema
* spec.validation.openAPIV3Schema: Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
+ (( retries += 1 ))
+ [[ 11 -gt 10 ]]
+ echo 'error: '\''kubectl apply -f k8s.ovn.org_vteps.yaml'\'' did not succeed, failing'
+ exit 1
error: 'kubectl apply -f k8s.ovn.org_vteps.yaml' did not succeed, failing
make: *** [Makefile:12: install-kind] Error 1
make: Leaving directory '/home/runner/work/ovn-kubernetes/ovn-kubernetes/test'
```
ah oops need to adjust my CEL validation cost. will take a look

tssurya commented on go-controller/pkg/crd/vtep/v1/apis/clientset/versioned/fake/clientset_generated.go:
```diff
@@ -0,0 +1,130 @@
+			return false, nil, err
+		}
+		return true, watch, nil
+	})
```
haha nice catch! will report it upstream :D

tssurya commented on docs/api-reference/userdefinednetwork-api-spec.md:
```diff
@@ -303,10 +353,13 @@ _Appears in:_
+| `transport` _[TransportOption](#transportoption)_ | Transport describes the transport technology for pod-to-pod traffic.<br />Allowed values are "NoOverlay" and "Geneve".<br />- "NoOverlay": The network operates in no-overlay mode.<br />- "Geneve": The network uses Geneve overlay.<br />When omitted, the default behaviour is Geneve. |  | Enum: [NoOverlay Geneve] <br /> |
+| `noOverlayOptions` _[NoOverlayOptions](#nooverlayoptions)_ | NoOverlayOptions contains configuration for no-overlay mode.<br />This is only allowed when Transport is "NoOverlay". |  |  |
+| `evpnConfiguration` _[EVPNConfiguration](#evpnconfiguration)_ | EVPNConfiguration contains configuration for EVPN mode.<br />This is only allowed when Transport is "EVPN". |  |  |
 
```
aaaah yea good catch dead rabbit! :) will fix it along with other review comments I get.

tssurya commented on go-controller/pkg/cni/types/types.go:
```diff
@@ -80,6 +80,15 @@ type NetConf struct {
+	// Transport describes the transport protocol for east-west traffic.
+	// Valid values are "nooverlay", "geneve", and "evpn".
+	// Defaults to "geneve".
+	Transport string `json:"transport,omitempty"`
```
@pliurh headsup this part might be conflicting with https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5780/commits/6698692414971346304c780b56753be68f9b4dcf but its fine, whichever PR lands first, the other one can rebase.

kyrtapz commented on go-controller/pkg/crd/userdefinednetwork/v1/cudn.go:
```diff
@@ -70,17 +76,22 @@ type NetworkSpec struct {
+	// EVPNConfiguration contains configuration for EVPN mode.
+	// This is only allowed when Transport is "EVPN".
+	// +optional
+	EVPNConfiguration *EVPNConfiguration `json:"evpnConfiguration,omitempty"`
```
I find `evpnConfiguration`/`noOverlayOptions` as field names in our API incorrect but I might be nitpicky.
I think all of the fields are configuration/option and we are not adding the suffix in fields like:
`Layer2Config` -> `layer2`
`Layer3Config` -> `layer3` 
`TransportOption` -> `transport`

I think I would consider `noOverlay` and `evpn` a better choice that is consistent with existing fields. 
I know this is just applying what's in the enhancement but I would like to discuss it before it is too late.

pliurh replied:
Yeah, I agree it makes sense to maintain consistency.

tssurya replied:
this is done.

pperiyasamy commented on go-controller/pkg/crd/userdefinednetwork/v1/evpn.go:
```diff
@@ -0,0 +1,61 @@
+	VTEP string `json:"vtep"`
+
+	// MACVRF contains the MAC-VRF configuration for Layer 2 EVPN.
+	// This field is required for Layer2 topology and forbidden for Layer3 topology.
```
I'm bit confused with this comment, workloads within a L3 CUDN across nodes still connect via MACVRF, right ? can you explain ?

tssurya replied:
my understanding is the following:

for L2 CUDN:

- MACVRF for east-west - cross node traffic (same node is just same switch in OVN level)
- IPVRF for north-south

for L3 CUDN:

- IPVRF for east-west - cross node traffic - since each node like Jaime said gets its own slice making the need for  a router necessary (same node is just same switch in OVN level)
- IPVRF for north-south

so with that in mind, the MACVRF is only allowed for L2 CUDNs.

for interconnecting CUDNs via EVPN i.e cross subnet routing I assume both connected network sides will use the IPVRF.

pperiyasamy replied:
ok, i guess even for L2 CUDN, each node gets its own subnet, but didn't know `ovn_cluster_router` was involved for connecting cross node traffic for L3 CUDN case, so in this scenario IPVRF makes sense.

pperiyasamy commented on go-controller/pkg/crd/userdefinednetwork/v1/evpn.go:
```diff
@@ -0,0 +1,61 @@
+
+// VRFConfig contains configuration for a VRF in EVPN.
+type VRFConfig struct {
+	// VNI is the VXLAN Network Identifier for this VRF.
```
should VNI be tied to VxLAN ? what if we plan to use some other overlay in future ?

tssurya replied:
I am +1 to making the fields generic
I guess we had briefly spoken about support for NVGRE or other dataplane in future but forgot to raise it in the OKEP ..

from briefly looking at the RFCs and asking AI to summarize them for me, this is what I found:
```
+------------------+---------------------+---------------+-------------------------+---------+
| Encapsulation    | Header Overhead     | Identifier    | Typical Use Case        | RFC     |
+------------------+---------------------+---------------+-------------------------+---------+
| MPLS             | 4 bytes/label       | 20-bit label  | WAN, SP networks, DCI   | 7432    |
+------------------+---------------------+---------------+-------------------------+---------+
| VXLAN            | 50 bytes (UDP+VXLAN)| 24-bit VNI    | Data center overlay     | 8365    |
+------------------+---------------------+---------------+-------------------------+---------+
| NVGRE            | 42 bytes (GRE)      | 24-bit VSID   | Microsoft Hyper-V       | 8365    |
+------------------+---------------------+---------------+-------------------------+---------+
| MPLS over GRE    | 24+ bytes           | 20-bit label  | IP-only underlay        | 8365    |
+------------------+---------------------+---------------+-------------------------+---------+
| PBB (802.1ah)    | 18 bytes            | 24-bit I-SID  | Large-scale L2VPN       | 7623    |
+------------------+---------------------+---------------+-------------------------+---------+
| GENEVE           | Variable            | 24-bit VNI    | Future/flexible         | (draft) |
+------------------+---------------------+---------------+-------------------------+---------+
```

I think going with Virtual Network Identifier is safer and well understood? 
and for the other fields like VRF and routetargets I assume they are all on BGP layer anyways so that's common for all of these.

tssurya replied:
> Yeah, VNI as Virtual Network Identifie is fine.
> 
> I guess our best bet with the current approach would be new transports, so "EVPN" is really "EVPN+VXLAN", whereas in the future we might have "EVPN+SRv6"... We would probably have a different EVPN structure types to hold the config for each... Different EVPN structure types could reuse common inner types.

yea I think based on dataplane this could vary cause some of the other dataplane options do seem similar to VXLAN so we could get away with what we have now and keep dataplane implementation agnostic like for NVGRE at least.. but yea if we have a need for SRv6 which needs more options then we'd probably need a new type of transport OR

> 
> The other approach would be to add an additional discriminator with additional nesting inside the EVPNConfig structure itself which could come up as something cleaner if we were more certain about what other dataplanes we might support in the future but a hard decision right now.

yea this ^ agreed its too far fetched to make any meaningful changes..

the only thing we could do now if we want to is change `transport: evpn` to `transport: evpn-vxlan` and same for the config type so that in future each evpn dataplane type is considered a seperate config type.. but its also easy to introduce new dataplane specific configs and mention in docs default is always vxlan or something like that. SO given we have a few options to deal with this, I think I'll resolve this comment and keep things simple and go with what we agreed for VNI's naming alone.

tssurya commented on go-controller/pkg/crd/vtep/v1/types.go:
```diff
@@ -0,0 +1,104 @@
+
+// VTEPStatus contains the observed state of the VTEP.
+type VTEPStatus struct {
+	// Conditions slice of condition objects indicating details about VTEP status.
```
good q, for now I went with the normal generic `Conditions` array.

there aren't many details on the BFD workings in the OKEP - but assuming this is handled by FRR and Type1 EVPN Route is supported, then:

Whoever does the status implementation could define different types in the condition array like we do for other features today, some examples:

1. `VTEPAssignmentFailure` - annotation missing case for whatever reason
2. `VTEPUnreachableLinkDown` OR
3. `Withdrawing Routes for UDN` --? this probably belongs in CUDN status more.
4. `Healthy` or `Success` or `Accepted` or whatever

jcaamano replied:
Hmmm

The BFD status is associated to a session between BGP peer routers, not to a VTEP. For the most part, I don't see a VTEP as something that can have a global dynamic BGP/EVPN up or down state.

A BGP session can be up or down (consequently a VTEP can be advertised or not) and in that context there are multiple multi-homing scenarios:

- Path redundancy: we have multiple peers on multiple paths so if one peer path goes down the other paths are still available.
- ECMP on L2 networks: If one node stops peering, workloads on L2 networks can still be reachable through other nodes. This ultimately wouldn't work: we can rely on other nodes for routing, but in contrast to the standard BGP case where we might have a functioning overlay independently from BGP, with EVPN we rely on the MACVRF which wouldn't work with no BGP session.
- EVPN Multi-homing: this is either assigning the same [ethernet segment ](https://docs.frrouting.org/en/latest/bgp.html#ethernet-segments)or same VTEP IP on multiple nodes. While the enhancement strictly mentions EVPN Multi-homing, there is no other explicit mention of this type of configuration so I am not really sure if we/I would be missing something here or if the enhancement is referring to something else, @trozet ?

Overall, I don't see an easy, adequate, or straight forward way to reflect this in the VTEP status.

tssurya commented on go-controller/pkg/crd/vtep/v1/types.go:
```diff
@@ -0,0 +1,104 @@
+// +kubebuilder:validation:MinItems=1
+// +kubebuilder:validation:MaxItems=2
+// +kubebuilder:validation:XValidation:rule="size(self) != 2 || !isCIDR(self[0]) || !isCIDR(self[1]) || cidr(self[0]).ip().family() != cidr(self[1]).ip().family()", message="When 2 CIDRs are set, they must be from different IP families"
+type DualStackCIDRs []CIDR
```
This is a good question :) 
I guess it wasn't captured in OKEP cause I don't see it..
but in dualstack I assume we would pick just the first cidr in the spec?
but on thinking more about this I feel this should just be allowing one CIDR - either v4 or v6 I guess.. I assume today ipsec tunnel endpoint works the same where we only pick one of the IPs?
I don't see us going to implement two tunnels or anything since we went with SVD approach.
I will update the VTEP CRD CEL to only allow for one cidr range...
nice catch @pperiyasamy

tssurya replied:
Was speaking with @trozet today a bit and he also mentioned a plausible v4 to v6 migration scenario where users might wanna have 2 tunnels for a brief interim period of a switch, so then having support for both CIDRs in the VTEP field might be useful if we plan to support that.

tssurya replied:
I tied the VTEPs strictly as the endpoint for underlay tunnel and forgot/didn't realise its also the nexthop value for the routes advertised over the underlay tunnels in the overlay... so keeping it dualstack aware is the right thing to do

jcaamano commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -705,6 +741,9 @@ type userDefinedNetInfo struct {
 	managementIPs       []net.IP
+
+	transport  string
+	evpnConfig *ovncnitypes.EVPNConfig
```
nit: I feel uneasy about using the actual `ovncnitypes.EVPNConfig` struct here which serves other purpose and might not be our best choice here, if not just now, likely in the future.

You could just store these flat, however a pointer to struct saves on memory :P

jcaamano replied:
`ovncnitypes.EVPNConfig` is a Netconf API type (convenient for external consumption) from which we translate into the multi-network types (convenient for internal consumption) so using the Netconf API type as an internal type starts to look like an anti-pattern to me.

It's fine though if its reuse is kept internal, however I wouldn't like the NetInfo interface to re-use a NetConf type and the current approach brings it one step closer to that. I would probably at least define an alias to it as a declaration of intentions that we don't want people importing `util` to use these multi-network types to also have to import the seemingly unrelated package `cni/types`.

Nothing against using a struct, I was just suggesting an alternate approach, my comment is addressing the use of the `ovncnitypes.EVPNConfig` type. UX is unaffected since the interface itself is flat. And discriminated union is not being used here since transport field is not cross-checked I guess.

This was just a nit after all, it is fine if it is just me seeing this looking weird.

tssurya commented on .github/workflows/test.yml:
```diff
@@ -491,6 +491,7 @@ jobs:
           - {"target": "traffic-flow-test-only","ha": "noHA", "gateway-mode": "shared", "ipfamily": "ipv4", "disable-snat-multiple-gws": "noSnatGW", "second-bridge": "1br", "ic": "ic-single-node-zones", "traffic-flow-tests": "1-24", "network-segmentation": "enable-network-segmentation"}
           - {"target": "tools", "ha": "noHA", "gateway-mode": "local", "ipfamily": "dualstack", "disable-snat-multiple-gws": "SnatGW", "second-bridge": "1br", "ic": "ic-single-node-zones", "network-segmentation": "enable-network-segmentation"}
           - {"target": "serial", "ha": "noHA", "gateway-mode": "shared", "ipfamily": "ipv4", "disable-snat-multiple-gws": "snatGW", "second-bridge": "1br", "ic": "ic-single-node-zones"}
+          - {"target": "evpn", "ha": "noHA", "gateway-mode": "local", "ipfamily": "dualstack", "disable-snat-multiple-gws": "noSnatGW", "second-bridge": "1br", "ic": "ic-single-node-zones"}
```
done.

mattedallo commented on go-controller/pkg/crd/vtep/v1/apis/applyconfiguration/vtep/v1/vtepallocation.go:
Is this stale?

Shall we consider cleaning up generated files before re-generating them in `update-codegen.sh`?

E.g. something like the following (within the CRD loop):

```
gen_dir="${SCRIPT_ROOT}/pkg/crd/${crd}/${api_version}/apis"
  rm -rf \
    "${gen_dir}/applyconfiguration" \
    "${gen_dir}/clientset" \
    "${gen_dir}/listers" \
    "${gen_dir}/informers"
```

mattedallo replied:
thanks for acknowledge it! both cases are good, with separate PR probably cleaner. 
but you know better than me how to proceed, so up to you  :)

jcaamano commented on go-controller/pkg/crd/userdefinednetwork/v1/evpn.go:
```diff
@@ -0,0 +1,61 @@
+// RouteTargetString represents a BGP route target in the format "<AS>:<VNI>".
+// +kubebuilder:validation:XValidation:rule="self.matches('^[0-9]+:[0-9]+$')", message="routeTarget must be in the format '<AS>:<VNI>'"
+// +kubebuilder:validation:XValidation:rule="int(self.split(':')[0]) >= 1 && int(self.split(':')[0]) <= 4294967295", message="AS number in routeTarget must be between 1 and 4294967295"
+// +kubebuilder:validation:XValidation:rule="int(self.split(':')[1]) >= 1 && int(self.split(':')[1]) <= 16777215", message="VNI in routeTarget must be between 1 and 16777215"
```
Supporting IP as the global administrator might just be straight forward, any specific reason not to support it from the get go?

tssurya replied:
> Supporting IP as the global administrator might just be straight forward, any specific reason not to support it from the get go?

https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5779#discussion_r2632868944

It wasn't called out in the OKEP phase, note that I picked up the types based on what was written in the enhancement. At this point same to what I wrote on the above thread, I don't mind either way - let's just decide on one: 1) it can be done later and API can be expanded if someone asks for it 2) do it now

on that thread it sounded like an OK thing to do for future because it wasn't thought of as part of design.

tssurya commented on test/scripts/e2e-cp.sh:
```diff
@@ -162,8 +162,10 @@ if [[ "${WHAT}" = "$SERIAL_LABEL" ]]; then
+EVPN_TESTS="EVPN"
 if [ "$ENABLE_ROUTE_ADVERTISEMENTS" != true ]; then
   skip $BGP_TESTS
+  skip $EVPN_TESTS
```
that's not the format all features are following so far.. I prefer coupling each feature into their own separation and being able to only turn off EVPN alone if need be.  But again none of these actually have developer docs so each of us are mostly doing our own way I guess.

jcaamano replied:
> I don't mind coupling it with BGP_TESTS

It's already coupled, because it is governed by the single `$ENABLE_ROUTE_ADVERTISEMENTS` above. 

You are making the split, but the split is not being used for anything. So my point of view would view would be add the split when we need it.

Not all features are equal, an in the case of EVPN, it depends on the general BGP being enabled. Not sure a single pattern fits all.

tssurya commented:
https://github.com/ovn-kubernetes/ovn-kubernetes/compare/8cbe2e79723688f5adcc8ee91b5a193f0eec9cfa..325017b647b89d40079fcd78699a3283aa31f8e3 is diff for new changes addressing review comments. I'll do a rebase now which can be ignored.

tssurya commented:
@trozet PTAL https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5779#discussion_r2634334229

jcaamano commented on test/scripts/e2e-cp.sh:
```diff
@@ -161,7 +161,7 @@ if [[ "${WHAT}" = "$SERIAL_LABEL" ]]; then
 fi
 
-BGP_TESTS="BGP"
+BGP_TESTS="BGP|EVPN"
```
> that BGP being on doesn't mean these new controllers will be on

but BGP being off does mean these (VTEP) controllers would be off, doesn't it? Which means they should also be tagged with the RouteAdvertisements feature?

tssurya replied:
ack let me do that change..

tssurya replied:
I've now done the change.. hope that satisfies your requirement.. if not lmk

jcaamano commented on go-controller/pkg/crd/userdefinednetwork/v1/evpn.go:
```diff
@@ -0,0 +1,95 @@
+// +kubebuilder:validation:XValidation:rule="self.startsWith('*:') || self.contains('.') || (int(self.split(':')[0]) <= 65535 && int(self.split(':')[1]) <= 4294967295) || (int(self.split(':')[0]) > 65535 && int(self.split(':')[1]) <= 65535)", message="for 2-byte AS (1-65535) local admin can be 1-4294967295; for 4-byte AS (65536-4294967295) local admin must be 1-65535"
+// +kubebuilder:validation:XValidation:rule="self.startsWith('*:') || !self.contains('.') || isIP(self.split(':')[0])", message="IPv4 address is invalid"
+// +kubebuilder:validation:XValidation:rule="self.startsWith('*:') || !self.contains('.') || (int(self.split(':')[1]) >= 1 && int(self.split(':')[1]) <= 65535)", message="local administrator for IPv4 format must be between 1 and 65535"
+// +kubebuilder:validation:XValidation:rule="!self.startsWith('*:') || int(self.split(':')[1]) <= 4294967295", message="local administrator for wildcard format must not exceed 4294967295"
```
BTW, FRR does take 0 both as the global and the local administrator so I probably wouldn't be too overzealous with it.

jcaamano replied:
> since every rule is evaluated even if the first rule fails AFAIU

I didn't think that was the case but I might be wrong. Do you have a reference for that?

jcaamano commented:
BTW, did you consider helm chart changes?

kyrtapz commented on test/e2e/testscenario/vtep/invalid-scenarios.go:
```diff
@@ -0,0 +1,101 @@
+
+import "github.com/ovn-org/ovn-kubernetes/test/e2e/testscenario"
+
+var Invalid = []testscenario.ValidateCRScenario{
```
Please note that the following comment is not directly related to the PR but I want to start a discussion using this as an example.

I think it would be better to run CRD validation tests as part of our unit tests as they shouldn't really require a cluster running. Doing a bit of research it should be fairly straightforward.
I would like to hear your thoughts on it and maybe we can open up a followup issue to change over the CRD validations, this would allow us for quicker testing and would remove the running cluster requirement.

kyrtapz commented on go-controller/pkg/cni/types/types.go:
```diff
@@ -80,6 +80,15 @@ type NetConf struct {
+
+	// EVPNConfig contains configuration for EVPN mode.
+	// Only valid when Transport is "evpn".
+	EVPNConfig *EVPNConfig `json:"evpnConfig,omitempty"`
```
shall we change the json tag to `evpn` to match the CRD?

tssurya commented on go-controller/pkg/crd/userdefinednetwork/v1/evpn.go:
```diff
@@ -0,0 +1,94 @@
+// See: https://docs.frrouting.org/en/stable-8.5/bgp.html#evpn-l3-route-targets
+//
+// +kubebuilder:validation:XValidation:rule="self.split(':').size() == 2",message="RT must contain exactly one colon"
+// +kubebuilder:validation:XValidation:rule="self.split(':').size() != 2 || (self.startsWith('*:') || self.split(':')[0].isIP() || self.split(':')[0].matches('[0-9]+'))",message="RT global administrator must be either '*', an IPv4 address, or a number"
```
@jcaamano it should be `isIP(self.split(':')[0])` changing this.

tssurya commented on go-controller/pkg/crd/vtep/v1/apis/clientset/versioned/fake/clientset_generated.go:
```diff
@@ -0,0 +1,130 @@
+	cs.AddReactor("*", "*", testing.ObjectReaction(o))
+	cs.AddWatchReactor("*", func(action testing.Action) (handled bool, ret watch.Interface, err error) {
+		var opts metav1.ListOptions
+		if watchActcion, ok := action.(testing.WatchActionImpl); ok {
```
@coderabbitai: I think we have agreed we can't do anything to fix this in this PR. See https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5779#discussion_r2675737824 you always seem to bring this up, 6th time now, learn to ignore things in life somethings for mental sanity reasons...- specially the not so important things we usually stress over for no reason..

tssurya commented:
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5779#discussion_r2675542632 is the last comment that's outstanding (I will push a fix for that).. if there are further comments, please drop there - I would like to have us merge this PR soon.

for rt6 support: https://www.rfc-editor.org/rfc/rfc5701 we can add it later if we plan to support it as well - should be CEL expansion hopefully and not considered a breaking change.

tssurya commented:
the latest push fixes the length and immutability aspects...
I'm going to also quickly rebase on master while at it and next push can be ignored..

mattedallo commented:
LGTM
Thank you @tssurya !

tssurya commented:
> > Two design decisions we took on the API PR that was not covered in OKEP:
> > 
> > 1. Keep VTEP CR immutable for now. Keep VTEP controller simple. In future we might need CIDR expansion but we will cross that bridge with a smaller update PR
> > 2. Keep VTEP cidr field as singlestack and expand later if we ever get to doing mutability and expansion and migration of dualstack etc etc we can expand it to dualstack
> 
> well hold that thought :P we are still debating..

nope ended up deciding in the opposite way.. mutable and dualstack vtep - we support everything and everything will have tests - anything claimed as supported via API with mutability will have proper test coverage and docs.

reverting the change to mutability.
